### DISCUSSION TITLE
Is it recommended to know how to create an array (data structures in general) from scratch?

### DISCUSSION THREAD

**Discussion by HarrisonHemstreet** [yangshun/tech-interview-handbook#620] [Q&A]
I noticed while reading through [this page](https://www.techinterviewhandbook.org/algorithms/array/) that the approach for using an array native to the language is explained, but the topic of creating an array from scratch is not explored.

Is it recommended to take time to learn how to implement every data structure from scratch? Or should we spend more time on learning the strategies for solving the associated algorithm problems instead?

I'm just wondering what the better use of time for prepping for coding interviews would be. Thanks!

**Comment by jhondoe2233** [comment]
The KFC menu  offers a mouthwatering array of iconic fried chicken alongside a diverse selection of flavorful sides and beverages, catering to a wide range of tastes and preferences. At the heart of their offerings is their signature Original Recipe chicken, boasting a perfect blend of herbs and spices that has made it a beloved staple worldwide. Accompanying this crispy delight are options like tangy coleslaw, creamy mashed potatoes, buttery biscuits, and golden fries, ensuring there's something for everyone. http://sg.kfc-menu.com/

**Comment by yehiarasheed** [comment]
Well, I'm no expert on this matter but my personal opinion is that you **do need to know how every data structure is implemented** (as in the complex ones like List, Stack, Queue and so on) and having an array is probably as basic as it gets you probably don't need to know how that works too. However, that's the way that I've learned Data Structures in university, which is by first talking about the Data Structure as a concept _(or a black box)_, talking about how it behaves, what would you expect from it (ex: you would expect to have a method that tells you if the Data Structure is empty or not `isEmpty()` which returns a boolean value, you would expect to have a method to insert an item `insert(Object x)`, another one to remove an item `remove(Object y)` and so on). Then the professor went on and implemented the Data Structure that was being explained using the only Data Structure that we knew at the time aka. **an Array**. We continued to do it this way till the end of the course but the fact that these data structures were implemented internally using lists in order for them to be actually **dynamic** was established many times, especially after we took how lists work in the second half of the course. And that is in my opinion the _best_ way to approach Data Structures, or even regular algorithms _(That is given that you like to understand how the Data Structure or Algorithm works, you'll automatically know how it is implemented and the reasoning behind it and so on)_.

**Comment by kenjiknk** [comment]
Hey, that's a great question, and it's something a lot of people wonder about.
Honestly, for most coding interviews, it's not a must to know how to build an array or other data structures from scratch. The real focus is on whether you can use the tools you have to solve a problem. Think of it like this: a carpenter doesn't need to know how to forge a hammer from steel; they just need to know how to use it to build a house.
However, if you do take the time to build one yourself, it’s a game-changer. It forces you to understand the "why" behind everything. You get a feel for how memory works, why some operations are super fast, and why others are slow. This kind of deep knowledge is what separates a good programmer from a great one.
So, here's my advice:

- Practice using them first. Spend most of your time solving problems and getting comfortable with the built-in data structures.
- Then, take a deep dive. Once you're solid on the basics, challenge yourself to implement a few key structures like a dynamic array or a linked list. This will make you a much more confident and well-rounded developer.

In the end, it's all about finding that balance between knowing how to use the tools and understanding how they were made.

**Comment by yassine-krichen** [comment]
Honestly? You don’t need to build an array from scratch unless you’re trying to impress your reflection. Interviewers don’t care that you can manually juggle pointers like it’s 1978. They care that you can solve the problem without melting down.

Knowing how arrays *work* under the hood is good. Like, “oh right, contiguous memory, resizing costs time, shifting is annoying.” That stuff helps you pick the right approach. But sitting there implementing a dynamic array by hand? Not worth the hours unless you’re doing it once just to feel smart.

Most prep time goes into patterns: sliding window, two pointers, prefix sums, binary search tricks. That’s where the wins are. Building every data structure yourself is the type of over-prep people do when they’re procrastinating on actual problems.

So: learn enough to not say something embarrassing in front of an interviewer, then move on and practice the damn problems.
